torchaudio/backend/utils.py

```diff
@@ -1,62 +1,74 @@
+"""Defines utilities for switching audio backends"""
 import warnings
-from typing import Any, Optional
+from typing import Optional, List
 
-from torchaudio._internal import module_utils as _mod_utils
-from . import soundfile_backend, sox_backend
+import torchaudio
+from torchaudio._internal.module_utils import is_module_available
+from . import (
+    no_backend,
+    sox_backend,
+    soundfile_backend,
+)
 
-_BACKEND = None
-_BACKENDS = {}
+__all__ = [
+    'list_audio_backends',
+    'get_audio_backend',
+    'set_audio_backend',
+]
 
 
-def list_audio_backends():
-    return list(_BACKENDS.keys())
+def list_audio_backends() -> List[str]:
+    """List available backends"""
+    backends = []
+    if is_module_available('soundfile'):
+        backends.append('soundfile')
+    if is_module_available('torchaudio._torchaudio'):
+        backends.append('sox')
+    return backends
 
 
-def set_audio_backend(backend: str) -> None:
-    """
-    Specifies the package used to load.
+def set_audio_backend(backend: Optional[str]) -> None:
+    """Set the backend for I/O operation
+
     Args:
         backend (str): Name of the backend. One of "sox" or "soundfile",
             based on availability of the system.
     """
-    if backend not in _BACKENDS:
+    if backend is not None and backend not in list_audio_backends():
         raise RuntimeError(
             f'Backend "{backend}" is not one of '
             f'available backends: {list_audio_backends()}.')
-    global _BACKEND
-    _BACKEND = backend
-
-
-def get_audio_backend() -> Optional[str]:
-    """
-    Gets the name of the package used to load.
-    """
-    return _BACKEND
 
+    if backend is None:
+        module = no_backend
+    elif backend == 'sox':
+        module = sox_backend
+    elif backend == 'soundfile':
+        module = soundfile_backend
+    else:
+        raise NotImplementedError(f'Unexpected backend "{backend}"')
 
-def _get_audio_backend_module() -> Any:
-    """
-    Gets the module backend to load.
-    """
-    if _BACKEND is None:
-        raise RuntimeError('Backend is not initialized.')
-    return _BACKENDS[_BACKEND]
+    for func in ['save', 'load', 'load_wav', 'info']:
+        setattr(torchaudio, func, getattr(module, func))
```

Comment:
Let's just make sure we make that conscious choice: `torchaudio.backend.__init__` invokes `utils._init_audio_backend()` which then modifies `torchaudio`.

Would it be advantageous to attach `load`, `save`, etc directly in `torchaudio.__init__` instead?

Reply:
I am not sure what you mean by `advantageous` here, but the current approach gives cleaner separation of `torchaudio.__init__` and `torchaudio.backend` because `backend` module provides 1. backend implementations, 2. way to manipulate backend, and 3. way to fetch information about current backend and available backends. This way, all the responsibilities on backend is closed within `torchaudio.backend` module and the only contract between `torchaudio.__init__` and `torchaudio.backend` is that `torchaudio.load/save/load_wav/info` attributes are reserved for `torchaudio.backend` module (implying that as long as `torchaudio.__init__` does not touch these attributes, `torchaudio.backend` understands everything about it), and these attributes are the only attributes outside of `torchaudio.backend` that `torchaudio.backend` module is allowed to touch.

If the initialization is moved to `torchaudio.__init__`, that means there are more contract between `torchaudio.__init__` and `torchaudio.backend`, like how the backend switch function works, and it requires `torchaudio.__init__` to know the internal mechanics of `torchaudio.backend`. This introduces more coupling between `torchaudio.__init__` and `torchaudio.backend` and adds maintenance cost.

Reply:
My main concern was the type of coupling that led to #712, when one module modifies another.

If I understand what you are saying: in order to avoid having the two depending on each other, we would need to bring `torchaudio.backend` completely within `torchaudio/__init__.py`. Is that what you are saying?

btw I am not advocating this :)

Reply:
> My main concern was the type of coupling that led to #712, when one module modifies another.

That's different. #712 was about new classes being defined inside of a module when a module function called from outside call.
Refactoring of backend switch is about achieving the proper [Separation on Concerns](https://en.wikipedia.org/wiki/Separation_of_concerns), by letting the `backend` module handle everything about backend, including initialization, so that main `torchaudio` module does not need to.

> If I understand what you are saying: in order to avoid having the two depending on each other, we would need to bring `torchaudio.backend` completely within `torchaudio/__init__.py`. Is that what you are saying?

No, that's totally opposite of what I am suggesting. It's better maitainable if `torchaudio/__init__.py` knows nothing about `backend` internal mechanism.

Reply:
Let me rephrase my concern. If I read `__init__`, I don't have any way of knowing that `load/save/etc` are being added to it since it's added by another module. Am I missing something?

How about setting the attribute to `torchaudio.backend` and simply importing in `torchaudio/__init__.py`? Thoughts?

Reply:
> Let me rephrase my concern. If I read `__init__`, I don't have any way of knowing that `load/save/etc` are being added to it since it's added by another module. Am I missing something?

Yes, that's the correct way to achieve the Separate of Concerns.

> How about setting the attribute to `torchaudio.backend` and simply importing in `torchaudio/__init__.py`? Thoughts?

No that way, the responsibility of backend switching is spread to `torchaudio/__init__.py` and `torchaudio.backend`, which introcudes more coupling. That's anti-pattern.

Just adding some comment about `save/load/info` assigned in `torchaudio.backend` would suffice.